### PR TITLE
[plugin.video.newson] 1.0.2

### PR DESCRIPTION
**Pull request by Lunatixz**
### Checklist:

- [x] My code follows the [add-on rules](http://kodi.wiki/view/Add-on_rules) and [piracy stance](http://kodi.wiki/view/Official:Forum_rules#Piracy_Policy) of this project. 
- [x] I have read the [CONTRIBUTING](https://github.com/xbmc/repo-plugins/blob/master/CONTRIBUTING.md) document
- [x] Each add-on submission should be a single commit with using the following style: [plugin.video.foo] v1.0.0

Additional information :
- Submitting your add-on to this specific branch makes it available to any Kodi version equal or higher than the branch name with the applicable Kodi dependencies limits.
- [add-on development](http://kodi.wiki/view/Add-on_development) wiki page.
- Kodi [pydocs](http://kodi.wiki/view/PyDocs) provide information about the Python API
- [PEP8](https://www.python.org/dev/peps/pep-0008/) codingstyle which is considered best practice but not mandatory.
- This add-on repository has automated code guideline check which could help you improve your coding. You can find the results of these check at [Codacy](https://www.codacy.com/app/Kodi/repo-plugins/dashboard). You can create your own account as well to continuously monitor your python coding before submitting to repo.
- Development questions can be asked in the [add-on development](http://forum.kodi.tv/forumdisplay.php?fid=26) section on the Kodi forum.

### CODE REVIEW

File: plugin.video.newson/default.py [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+        #using xmltodict, feedparser too bulky for the job.
+        feed = feedparser.parse(url)
+        for item in feed['entries']:
+            if item and 'summary_detail' in item:
+                for vids in item['media_content']:
+                    title = item['title']
+                    url   = vids['url']
+                    plot  = item['summary']
+                    thumb = item['media_thumbnail'][0]['url']
+                    infoLabels ={"label":title ,"title":title,"plot":plot}
```
Comment (learningit) [inline plugin.video.newson/default.py]: Please set mediatype in infolabels

File: plugin.video.newson/default.py [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+            return
+        for channel in data:
+            if state in channel['config']['state']:
+                chid   = channel['identifier']
+                title  = channel['title']
+                icon   = (channel['icon'] or ICON)
+                vidURL = channel['config']['localvodfeed']
+                if vidURL:
+                    label      = "%s - %s" % (chid, title)
+                    infoLabels ={"label":label ,"title":label}
```
Comment (learningit) [inline plugin.video.newson/default.py]: Please set mediatype in infolabels

File: plugin.video.newson/default.py [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+                    #multiple urls, only and unique.
+                    url = (stream['Url'].split('.m3u8?')[0])+'.m3u8'
+                    offset = stream['OffsetFromNow']
+                    delay  = url+'&delay=%d'
+                    #todo do something with delay option?
+                    if url not in urls:
+                        urls.append(url)
+                        chid = chid+'.%d'%idx if idx > 0 else chid
+                        label      = "%s - %s" % (chid, title)
+                        infoLabels ={"label":label ,"title":label}
```
Comment (learningit) [inline plugin.video.newson/default.py]: Please set mediatype in infolabels

File: plugin.video.newson/default.py [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+                        infoArt    ={"thumb":icon,"poster":icon,"fanart":FANART,"icon":icon,"logo":icon} 
+                        self.addLink(title, url, 9, infoLabels, infoArt)
+        del urls[:]
```
Comment (learningit) [inline plugin.video.newson/default.py]: Why is this necessary? I would expect urls[] to be auto-deleted at the end of the method.

File: plugin.video.newson/default.py [hunk 5]
```diff
@@ -0,0 +1,270 @@
...
+        
+    def openURL(self, url):
+        try:
+            cacheResponce = self.cache.get(ADDON_NAME + '.openURL, url = %s'%url)
+            if not cacheResponce:
+                request = urllib2.Request(url)
+                request.add_header('User-Agent','Mozilla/5.0 (Windows; U; MSIE 9.0; Windows NT 9.0; en-US)')
+                page = urllib2.urlopen(request, timeout = TIMEOUT)
+                log(page.headers['content-type'])
+                results = json.load(page)
```
Comment (learningit) [inline plugin.video.newson/default.py]: Just a curiosity piece here: why wouldn't you check for a 'Content_Encoding' == 'gzip' and do the decompress if so?
Are you ensured that the source site and gateways won't compress? In my experience even not specifying gzip in 'Accept-Encoding' doesn't mean that you won't get a zipped response (most gateways do what they want, not what you tell them).

File: plugin.video.newson/default.py [hunk 6]
```diff
@@ -0,0 +1,270 @@
...
+                page.close()
+                self.cache.set(ADDON_NAME + '.openURL, url = %s'%url, results, expiration=datetime.timedelta(hours=1))
+            return self.cache.get(ADDON_NAME + '.openURL, url = %s'%url)
+        except urllib2.URLError, e:
+            log("openURL Failed! " + str(e), xbmc.LOGERROR)
+        except socket.timeout, e:
+            log("openURL Failed! " + str(e), xbmc.LOGERROR)
+        except:            
+            xbmcgui.Dialog().notification(ADDON_NAME, 'Site unreachable, try again later...', ICON, 4000)
```
Comment (learningit) [inline plugin.video.newson/default.py]: Please use translatable string in strings.po.

File: plugin.video.newson/default.py [hunk 7]
```diff
@@ -0,0 +1,270 @@
...
+        stateLST  = []
+        data = self.openURL(BASE_API)
+        if not data:
+            return []
+        for channel in data:
+            state.append(channel['config']['state'])
+        states = collections.Counter(state)
+        for key, value in sorted(states.iteritems()):
+            stateLST.append(("%s"%(key), key , '{}'))
+        del state[:]
```
Comment (learningit) [inline plugin.video.newson/default.py]: Also don't think this del is necessary, python should do it for you.

File: plugin.video.newson/default.py [hunk 8]
```diff
@@ -0,0 +1,270 @@
...
+            msg += ' ,' + traceback.format_exc()
+        xbmc.log(ADDON_ID + '-' + ADDON_VERSION + '-' + stringify(msg), level)
+        
+def stringify(string):
+    if isinstance(string, list):
+        string = stringify(string[0])
+    elif isinstance(string, (int, float, long, complex, bool)):
+        string = str(string) 
+    if isinstance(string, basestring):
+        if not isinstance(string, unicode):
```
Comment (learningit) [inline plugin.video.newson/default.py]: The logic in this if-elif-else is pretty funky. You might want to look at it again.

Comment (Lunatixz) [inline plugin.video.newson/default.py]: it's not an if elif else statement. Stringify doesn't cover all possible types(). The if elif attempts to convert to a string. Then it checks to make sure only strings are being evaluated for encoding.

File: plugin.video.newson/default.py (hunk 5 above)
Comment (Lunatixz) [inline plugin.video.newson/default.py]: I've been periodically checking their return headers for changes. I believe the site is in the middle of making changes. I'll check again for compression.

File: plugin.video.newson/default.py (hunk 7 above)
Comment (Lunatixz) [inline plugin.video.newson/default.py]: Probably not necessary, I'll remove it.

File: plugin.video.newson/default.py (hunk 3 above)
Comment (Lunatixz) [inline plugin.video.newson/default.py]: Doesn't that set the mediatype to video?
https://github.com/Lunatixz/repo-plugins/blob/a98069e3f93d237290eef3e50b78ac3d3f53420f/plugin.video.newson/default.py#L214

Comment (Lunatixz) [inline plugin.video.newson/default.py]: ic there is another mediatype, was a bit confused. 
https://forum.kodi.tv/showthread.php?tid=250936&pid=2319280#pid2319280

File: plugin.video.newson/default.py (hunk 8 above)
Comment (learningit) [inline plugin.video.newson/default.py]: Yeah I get that. It's not a show stopper, but the logic is messy.

File: plugin.video.newson/default.py (hunk 3 above)
Comment (learningit) [inline plugin.video.newson/default.py]: https://forum.kodi.tv/showthread.php?tid=250936&pid=2319280#pid2319280
is the one I'm referring to.

set infolabel "mediatype" to one of these:
""
"music"
"artist"
"album"
"song"
"video"
"set"
"musicvideo"
"movie"
"tvshow"
"season"
"episode"

File: plugin.video.newson/default.py (hunk 5 above)
Comment (learningit) [inline plugin.video.newson/default.py]: Why not just add something like this and be done with it. It's a possible problem waiting to happen:

      response = urllib2.urlopen(req, timeout=TIMEOUT)
      page = response.read()
      if response.info().getheader('Content-Encoding') == 'gzip':
         self.log("Content Encoding == gzip")
         page = zlib.decompress(page, zlib.MAX_WBITS + 16)
      results = json.load(page)

File: plugin.video.newson/default.py (hunk 8 above)
Comment (learningit) [inline plugin.video.newson/default.py]: ok.

File: plugin.video.newson/addon.xml [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+    </requires>
+    <extension point="xbmc.python.pluginsource" library="default.py">
+        <provides>video</provides>
+    </extension>
+    <extension point="xbmc.addon.metadata">
+        <summary lang="en_GB">Local News Nationwide</summary>
+        <description lang="en_GB">NewsON provides instant access to live or on-demand broadcasts from over 170 local stations in 113 U.S. markets. Access previous newscasts for up to 48 hours after they air, and share interesting stories with the integrated social media share function directly from the app. Search for a local station with the interactive map or watch the station closest to you on smartphones, tablets, or Roku streaming players. Save favorite stations, and stay updated with breaking news alerts sent directly to mobile devices from the local source of the news. Stay connected anytime, anywhere without a cable subscription or login.</description>
+        <platform>all</platform>
+        <license>GNU GENERAL PUBLIC LICENSE. Version 3, June 2007</license>
+        <source>https://github.com/Lunatixz/XBMC_Addons/tree/master/plugin.video.newson</source>
```
Comment (learningit) [inline plugin.video.newson/addon.xml]: Ah, now I understand why your commits always show full diffs. You haven't forked the xbmc/repo-plugins.
You just do a pull request against your addon git branch. If you did the fork and create a separate branch for each version, as you add commits and squash them you would only get a small diff which is very easy to review.

Comment (Lunatixz) [inline plugin.video.newson/addon.xml]: hmm, I do forked xbmc/repo-plugins and create a new branch for each push version using krypton as a base... I then delete the branch once merged. Is that incorrect? Does the source of the addon.xml matter? Thought it was reference info only...

Comment (learningit) [inline plugin.video.newson/addon.xml]: you're right addon.xml doesn't matter. I've been trying to figure out why there's a full diff for everything you do. Normally I see just the changes which is much easier to review. I thought that just doing a compare against your git without a up to date forked version might be the reason.

**Comment by learningit** [comment]
When you add a new commit please name it properly as well. "recommended changes" is not correct.

**Comment by Lunatixz** [comment]
@enen92 is a master at merging... On my side I'd have to figure out how to rename a commit... That may take some time. My github skills stink, Most likely i'll end up making a mess.

**Comment by learningit** [comment]
Ok. Someone else needs to review your PRs from now on.

**Comment by Lunatixz** [comment]
lol, I welcome your reviews... I'm trying to limit my commits to a single push. In this case we are pushing my knowledge of GitHub... Which will require sometime for me to learn new git shell commands. Sorry :(

**Comment by Lunatixz** [comment]
@learningit Not sure if I made things worse? I tried to amend the commit name...
